### PR TITLE
feat(api): wire up TC deactivate, and add module cmd exec endpoint

### LINKED ISSUE #2981: Thermocycler: Driver: Pause/ deactivate module
## overview

As a robot user, I would like to have a TC driver to be able to communicate the following with the TC via gcodes:
- [ ] pause the TC
- [ ] deactivate TC

## Open question
What does it mean to pause a thermocycler? Stay on its current step or stop doing anything. How does that differentiate the action from deactivate?

@pantslakz does this pause the countdown, and then remainder of countdown runs after un-pause? Or does the countdown continue but pause would keep it from moving on to the next cycle?

### PR DESCRIPTION
**Pull request by b-cooper**
## Overview

Allow users to deactivate TC from context/HC. This does not add pause and resume functionality as it
would be largely redundant given the existing api using a combination of tc.set_temperature,
tc.wait_for_temp, and robot.pause. This adds a new endpoint that creates an HTTP interface for
effecting module settings. The aim is to relieve the main UX use case of the planned 'pause'
function by giving users an avenue to set module attributes before a protocol run is initiated.

Closes #2981

## Changelog

- tc.deactivate() now hits the TC board
- tc driver uses GCODE response parsing utils to remove chances of TypeError/ValueError
- New endpoint allows control of module:
   ```
    POST to /modules/{mod_serial}
    body = {
       "command_type": "set_temperature",
       "args": [4]
    }
    ```

where `"command_type"` must correspond to a HC instance method for that particular module class,
and  `"args"` must correspond to an array of parameters that function expects. If it is a side effect function with 0 params, either passing and empty array or omitting the `"args"` key should work.

### CODE REVIEW

File: api/src/opentrons/server/endpoints/control.py [hunk 1]
```diff
@@ -163,6 +163,52 @@ def hw_from_req(req):
...
+    data = await request.json()
+    command_type = data.get('command_type')
+    args = data.get('args')
+
+    if ff.use_protocol_api_v2():
+        hw_mods = await hw.discover_modules()
+    else:
+        hw_mods = hw.attached_modules.values()
+
+    if len(hw_mods) > 0:
```
Comment (sfoster1) [inline api/src/opentrons/server/endpoints/control.py]: [quibble] There's a lot of nested logic here. Could be done sequentially instead and be a bit clearer, at least to me

File: api/src/opentrons/drivers/thermocycler/driver.py [hunk 2]
```diff
@@ -153,7 +152,7 @@ def _wait_for_ack(self):
         self._send_command(SERIAL_ACK, timeout=DEFAULT_TC_TIMEOUT)
 
     def _send_command(self, command, timeout=DEFAULT_TC_TIMEOUT):
-        command_line = command + ' ' + TC_COMMAND_TERMINATOR
+        command_line = command + ' ' + SERIAL_ACK
```
Comment (Laura-Danielle) [inline api/src/opentrons/drivers/thermocycler/driver.py]: Pls change back to `TC_COMMAND_TERMINATOR`

File: api/src/opentrons/server/endpoints/control.py [hunk 3]
```diff
@@ -163,6 +163,49 @@ def hw_from_req(req):
...
+    command_type = data.get('command_type')
+    args = data.get('args')
+
+    if ff.use_protocol_api_v2():
+        hw_mods = await hw.discover_modules()
+    else:
+        hw_mods = hw.attached_modules.values()
+
+    if len(hw_mods) == 0:
+        return web.json_response({"message": "No connected modules"})
```
Comment (sanni-t) [inline api/src/opentrons/server/endpoints/control.py]: status 404 too?

**Comment by sanni-t** [comment]
Just that one comment. Rest all looks great
🤘

**Pull request merged**